File: zemiao/257.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        ans = []
        
        def dfs(temp, node):
            temp += str(node.val) if temp=="" else "->"+str(node.val)
            if not node.left and not node.right:
                ans.append(temp)
            
            if node.left:
                dfs(temp, node.left)
            if node.right:
                dfs(temp, node.right)
        
        if not root:
            return ans
        dfs("",root)
        return ans

File: chunlin/257.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        helper(root, result, String.valueOf(root.val));
        return result;
        
    }
    private void helper(TreeNode root, List<String> result, String cur) {
        if (root.left == null && root.right == null) {
            result.add(cur);
        }
        if (root.left != null) {
            helper(root.left, result, cur + "->" + String.valueOf(root.left.val));   
        }
        if (root.right != null) {
            helper(root.right, result, cur + "->" + String.valueOf(root.right.val));
        }
    }
}
